cibuildwheel/macos.py

```diff
@@ -641,11 +640,17 @@ def build(options: Options, tmp_path: Path) -> None:
 
             # we're all done here; move it to output (overwrite existing)
             if compatible_wheel is None:
-                with contextlib.suppress(FileNotFoundError):
-                    (build_options.output_dir / repaired_wheel.name).unlink()
+                output_dir = build_options.output_dir.resolve()
+                output_wheel = output_dir.joinpath(repaired_wheel.name).resolve()
+                output_wheel.unlink(missing_ok=True)
 
-                shutil.move(str(repaired_wheel), build_options.output_dir)
-                built_wheels.append(build_options.output_dir / repaired_wheel.name)
+                # shutil.move() will rename the file to what we were expecting to be the parent directory if we don't ensure it exists
+                output_dir.mkdir(parents=True, exist_ok=True)
+
+                # using shutil.move() as Path.rename() is not guaranteed to work across filesystem boundaries
+                # explicit str() needed for Python 3.8
+                shutil.move(str(repaired_wheel), str(output_wheel))
+                built_wheels.append(output_wheel)
```

Comment:
IMHO, we should add a `move` function  in `util.py` which takes care of unlinking, creating the parent directory and actually moving the wheel. It could be reused on all 3 platforms (workaround might be needed for local builds on macOS with pyodide for exemple).

Reply:
Hi mayeut,
That's a good idea.
Also a good catch: I missed updating pyodide as I wasn't building for it on my test project.

I'd be happy to put the change together later today.

What's the issue that you would see with local macos-pyodide builds?

Reply:
I'd imagine the util function has a signature along the lines of: `movewheel(wheel: Path) -> output_wheel: Path`, directly accesses `build_options` as this is global but leaves `built_wheels.append` at the calling site for clarity.

Does that make sense?

Reply:
The function signature should be something like `move(src: Path, dst: Path) -> None`

Reply:
How about this? (See latest push). I'm happy with the shape of the function and it follows shutil.move closely (which also returns the destination path).

I'm slightly less happy with the import inside the function but felt that logging the move was valuable enough to pay that price.

I've [tested against macos & windows](https://github.com/MusicalNinjaDad/cibuildwheel/actions/runs/9403124615). I couldn't test pyodide as the test project I'm using is rust-based and I haven't got pyodide/emscripten working for that yet ...

Reply:
What sort of issues are you hitting? It should Just Work™ in most cases, so there's a chance it's hitting a Pyodide or cibuildwheel bug.

Reply:
We can probably split utils up later to avoid the circular import, it's pretty big. Not something to worry about for this PR.

Why is the warning at every usage? Would it make sense to warn inside the function if it doesn't copy where it's supposed to?

Reply:
> What sort of issues are you hitting? It should Just Work™ in most cases, so there's a chance it's hitting a Pyodide or cibuildwheel bug.

rust seemed to be targetting the wrong target on the one time I tried "just switching it on" as part of testing this PR. Looking at the code and the implementation PR on cibuildwheel side, I get the impression that I "just" need to add a few config entries into pyproject.toml, find the best way to build in pipeline, check whether ADO artefact repos can accept the wheels, and set up some kind of test environment to play with the result...

It looks interesting enough that I'm likely to try playing with it in my sandbox project soon

Reply:
> Why is the warning at every usage? Would it make sense to warn inside the function if it doesn't copy where it's supposed to?

I'd consider it to be in the responsibility of the main script to double check that a called util did what was expected, that also covers the risk of later changes to utils etc.

Adding the warning to `util.move_file` feels like a no-op to me ... the only line that would effectively be checked is whether `shutil.move` did what was expected (?)